### PR TITLE
peroxide: init at 0.5.0

### PR DESCRIPTION
**Pull request by aidalgol**
###### Description of changes
Add NixOS service for [peroxide](https://github.com/ljanyst/peroxide), a fork of the official [ProtonMail bridge](https://github.com/ProtonMail/proton-bridge) that tries to be similar to the third-party [Hydroxide](https://github.com/emersion/hydroxide) while reusing as much code from the official bridge as possible.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/misc/ids.nix
Comment (SuperSandro2000) [inline nixos/modules/misc/ids.nix:357]: Please use dynamic user.

File: pkgs/applications/networking/peroxide/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/peroxide/default.nix:44]: This goes before meta. We should probably correct it from where it was copied, too.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:166]: Do these even matter? We could probably default them to nixos-user or something.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:147]:
```suggestion
    environment.etc."peroxide.conf".source = settingsFormat.generate "peroxide-config.yaml" cfg.settings;
```

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:136]:
```suggestion
        if [[ ! -e ${cfg.stateDir}/key.pem && ! -e ${cfg.stateDir}/cert.pem ]]; then
```

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:140]:
```suggestion
          (
            cd ${cfg.stateDir} || exit
            ${pkgs.peroxide}/bin/peroxide-cfg -action gen-x509 \
              -x509-org ${cfg.selfSignedCert.organization} \
              -x509-cn ${cfg.selfSignedCert.hostname}
          )
```

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:130]: The package should be configurable and rather be in path.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:83]: I don't see the usecase for this option tbh.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:20]: If we are using StateDir in systemd this must be under /var/lib. I am also not sure if this needs to be configurable at all.

File: nixos/modules/misc/ids.nix
Comment (aidalgol) [reply nixos/modules/misc/ids.nix:357]: How do I do that?

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:167]: They do have default values, so this assert is a bit overkill.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:141]: I'm not sure about this.  It will suppress any error from `cd` and exit early with a "success" return code, which is not what we want, is it?

File: pkgs/applications/networking/peroxide/default.nix
Comment (aidalgol) [reply pkgs/applications/networking/peroxide/default.nix:44]: I did not copy it from anywhere; I was just unaware of this style convention.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:20]: Searching for `stateDir` on search.nixos.org/options returns only a single page worth of results, so making it configurable does not appear to be the norm.  I'm happy to move it to a `let` binding.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:84]: Me neither, now that you mentioned it.  We still need to set the `CacheDir` peroxide setting because the default won't work on NixOS, but I have moved it to outside the user-configurable options.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:131]: OK, I'll make it a module option, but I'm not sure what you mean by "and rather be in path."

File: nixos/modules/misc/ids.nix
Comment (aidalgol) [reply nixos/modules/misc/ids.nix:357]: Done

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:109]: All that documentation/explanation written above just because of this option. I think that means you shouldn't use this option.

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:33]: If the user changes this option and rebuilds their system what happens? Something entirely not reproducible.

This option has questionable value. I suggest we just set a hard coded value and if the user needs something different they are free to provide their own certificate.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:109]: I initially did not, but @SuperSandro2000 requested changing to use `DynamicUser` in their review.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix]: Done.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [reply nixos/modules/services/networking/peroxide.nix:82]: If it does not easily work, then we can't use it but we definitely don't need to hardcode an ID. That can be dynamically allocated.

Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:82]: What is the correct way to dynamically allocate a system user (ID)?

Comment (aanderse) [reply nixos/modules/services/networking/peroxide.nix:82]: `users.users.foo.isSystemUser`

**Comment by SuperSandro2000** [comment]
Please rebase

**Comment by aidalgol** [comment]
Rebased against recent `master` and updated to latest upstream release.

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:86]: This won't work because `$CONFIGURATION_DIRECTORY` is owned by `root`. I suggest using `StateDirectory` and `StateDirectoryMode` here instead and only provision the credentials if not defined in `cfg.settings`.

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:78]: You're missing [log rotation](https://github.com/ljanyst/peroxide/blob/master/peroxide.logrotate).

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:68]: Adding a `description` would be nice.

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [reply nixos/modules/services/networking/peroxide.nix:78]: Upstream [suggestion](https://github.com/ljanyst/peroxide/blob/master/install.sh):
```suggestion
        LogsDirectory = "peroxide";
        LogsDirectoryMode = "0750";
```

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:81]: Upstream [suggestion](https://github.com/ljanyst/peroxide/blob/master/install.sh):
```suggestion
        CacheDirectory = "peroxide peroxide/cache";
        CacheDirectoryMode = "0700";
```

File: nixos/modules/services/networking/peroxide.nix
Comment (aanderse) [inline nixos/modules/services/networking/peroxide.nix:99]:
```suggestion
    environment.etc."peroxide.conf".source = settingsFormat.generate "peroxide.conf" cfg.settings;
```

**Review by aanderse — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix:78]: > You're missing [log rotation](https://github.com/ljanyst/peroxide/blob/master/peroxide.logrotate).

How should that be set in a NixOS module?

**Comment by aanderse** [comment]
I made a few changes to the module and pushed them [here](https://github.com/NixOS/nixpkgs/commit/718330473c125f40831a5833bdee92e37835384b). Please either add this commit to yours or squash it together.

I did some testing both using the self signed certificates as well as some Let's Encrypt provisioned ones, and then tested this against `go` and `dotnet` applications without any trouble.

The only missing piece is to modify the description of the `enable` option to include either a link to [these](https://github.com/ljanyst/peroxide#user-management) instructions, or the instructions themselves.

I'm really hoping `peroxide` will pan out for me because I'm at my wits end with `protonmail`... Both `hydroxide` and `protonmail-bridge` keep breaking for me and don't have great user experience. :face_exhaling: Anywho, I'll sign up to maintain this module with you in hopes that this is a long term solution for me.

**Comment by aidalgol** [comment]
> I made a few changes to the module and pushed them [here](https://github.com/NixOS/nixpkgs/commit/718330473c125f40831a5833bdee92e37835384b). Please either add this commit to yours or squash it together.

I have manually combined your changes from that commit into the pending changes I had locally.  A few changes I did not bring over:
- Renaming the let binding `settingsFormat` to `format`.
  When I see just `format` I think of format strings, so I find the current name clearer.
- `environment.systemPackages = [ cfg.package ];`
  This does not seem to be necessary, or a convention that other NixOS modules follow.
- Code rearranging.


Once I have your approval, I will rebase against current `master`.

**Review by aanderse — CHANGES_REQUESTED** [review]
Awesome! Yeah a few of my changes weren't important so all good 👍

The `environment.systemPackages` change is needed, though. Upstream documentation that we link to won't with without it. And as far as I understand proton will drop sessions after 60 days so every 60 days you'll need to run `sudo -u peroxide peroxide-cfg ...`. And of course the first time you run the application. So please add the package back to the system.

Also i had some issues with some systems connecting and forgot to do this but a `logLevel` option of type `enum` would be a nice feature if you're interested in adding.

Thanks so much! Looking good. We'll get this merged real soon.

**Comment by aanderse** [comment]
Don't hesitate to ask if you need any pointers. I'm keen to get this merged. I'd like to backport this to `22.11` as well, so consider putting the release notes in there if you don't mind.

**Comment by aidalgol** [comment]
I'm actually running NixOS 22.11, and with the latest revision of this branch, I am now getting an error when I go to rebuild my system config:
```console
error: The option `systemd.services.peroxide.services' does not exist. Definition values:
       - In `/nix/store/cbgsf0ry54nss4gvsj8ynim3sx90gr6c-source/nixos/modules/services/networking/peroxide.nix':
           {
             logrotate = {
               settings = {
                 peroxide = {
                   compress = true;
```

This is the contents of my system `flake.nix`:
<details>
<summary>flake.nix</summary>

```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.11";
    nixpkgs-peroxide.url = "git+file:///home/aidan/src/nixpkgs?ref=peroxide";
  };
  outputs = { self,
              nixpkgs,
              nixpkgs-peroxide }: {
    nixosConfigurations.aidaltower = let
      system = "x86_64-linux";
    in nixpkgs.lib.nixosSystem {
      inherit system;
      specialArgs = {
        inherit
          nixpkgs-peroxide;
        peroxide-module = "${nixpkgs-peroxide}/nixos/modules/services/networking/peroxide.nix";
      };
      modules = [
        ./configuration.nix
      ];
    };
  };
}
```
</details>

Then in `configuration.nix`,
```nix
# ...

  imports = [
    peroxide-module
    ./hardware-configuration.nix
    # ...
  ];

# ...

  # Run peroxide ProtonMail bridge.
  services.peroxide = {
    enable = true;
    package = nixpkgs-peroxide.legacyPackages.x86_64-linux.peroxide;
    logLevel = "Info";
  };

# ...
```

**Review by aanderse — APPROVED** [review]
I did a full test of the latest push to be sure and everything is working great :+1:

@SuperSandro2000 anything else before we merge?

**Comment by SuperSandro2000** [comment]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

**Comment by aidalgol** [comment]
> Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

I have changed the commit message to reflect that it is a NixOS module.

**Comment by SuperSandro2000** [comment]
I was more after the point that it should be one commit for the package and one for the module including the changelog entry.

**Comment by aidalgol** [comment]
OK, done.

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:21]: This should use mkEnableOption

File: nixos/modules/services/networking/peroxide.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/peroxide.nix:23]: and this mkPackageOption

Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix]: Done.

File: nixos/modules/services/networking/peroxide.nix
Comment (aidalgol) [reply nixos/modules/services/networking/peroxide.nix]: Done.

**Comment by aidalgol** [comment]
Rebased onto current `master`.

**Comment by aidalgol** [comment]
@aanderse I've moved the release notes to 23.03 since that release is now only a month away, so I don't think it's worth the effort to backport to 22.11 at this point.

**Comment by aanderse** [comment]
Are we good to merge this everyone? I think we should.

**Comment by emptyflask** [comment]
> @aanderse I've moved the release notes to 23.03 since that release is now only a month away, so I don't think it's worth the effort to backport to 22.11 at this point.

BTW looks like there's no 23.03 release, just 23.05

**Comment by aidalgol** [comment]
> > I've moved the release notes to 23.03 since that release is now only a month away, so I don't think it's worth the effort to backport to 22.11 at this point.
> 
> BTW looks like there's no 23.03 release, just 23.05

Right, of course.  I somehow thought the release was in March when I typed that.

**Comment by aidalgol** [comment]
Just force-pushed to fix an incorrect hash.  No idea why that didn't fail the ofborg run.

**Comment by aidalgol** [comment]
Rebased onto current `master`.

**Pull request merged**
